File: src/core/response-function.ts
import { AccessDeniedError, PageUnavailableError, RateLimitedError, ServerSideError, SessionExpiredError, SuspendedIPError, type SessionHandle } from "~/models";
import forge from "node-forge";
import { AES } from "../api/private/aes";
import pako from "pako";
import { aesKeys } from "../api/private/keys";

export class ResponseFN {
  public constructor (
    private session: SessionHandle,
    public data: any
  ) {
    this.session.information.order++;
    const content = data;

    try {
      const response = JSON.parse(content);
      if (response.Erreur) {
        const error = response.Erreur.Titre || "Server Error";
        throw new ServerSideError(error);
      }

      this.data = response.donneesSec;

      if (!this.session.information.skipEncryption) {
        this.decrypt();
      }

      if (!this.session.information.skipCompression) {
        this.decompress();
      }

      if (typeof this.data === "string") {
        this.data = JSON.parse(this.data);
      }

      if (typeof this.data?._Signature_?.Erreur !== "undefined") {
        throw new ServerSideError(this.data._Signature_.MessageErreur);
      }
    }
    catch (error) {
      if (content.includes("La page a expir")) {
        throw new SessionExpiredError();
      }

      else if (content.includes("Votre adresse IP ")) {
        throw new SuspendedIPError();
      }

      else if (content.includes("La page dem") || content.includes("Impossible d'a")) {
        throw new PageUnavailableError();
      }

      else if (content.includes("Vous avez d")) {
        throw new RateLimitedError();
      }

      // "Accès refusé", we just trying to prevent using accents.
      else if (content.includes("s refus")) {
        throw new AccessDeniedError();
      }

      throw error;
    }
  }

  private decrypt (): void {
    const { iv, key } = aesKeys(this.session);

    this.data = AES.decrypt(this.data, key, iv);

    if (!this.session.information.skipCompression) {
      this.data = forge.util.bytesToHex(this.data);
    }
  }

  private decompress (): void {
    const bytes = forge.util.hexToBytes(this.data);
    const compressed = new Uint8Array(Array.from(bytes).map((char) => char.charCodeAt(0)));
    this.data = pako.inflateRaw(compressed, { to: "string" });
  }
}

File: library/src/jvmMain/kotlin/core/response-function.kt
package core

import api.private.AES
import api.private.aesKeys
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import models.SessionInformation
import models.errors.*
import okio.ByteString.Companion.decodeHex
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.util.zip.Inflater

actual class ResponseFN actual constructor(
    private val sessionInfo: SessionInformation,
    actual var data: String,
) {
    actual var jsonData: JsonElement

    init {
        this.sessionInfo.order++
        val content = data

        try {
            this.data = Json.encodeToString(Json.parseToJsonElement(data).jsonObject["donneesSec"]!!)

            if (!this.sessionInfo.skipEncryption) {
                this.decrypt()
            }

            if (!this.sessionInfo.skipCompression) {
                this.decompress()
            }

            this.jsonData = Json.parseToJsonElement(this.data)

            if (this.jsonData.jsonObject.containsKey("_Signature_")
                && this.jsonData.jsonObject["_Signature_"]!!.jsonObject["Erreur"]?.jsonPrimitive?.boolean == true
            )
                throw ServerSideError(this.jsonData.jsonObject["_Signature_"]!!.jsonObject["MessageErreur"]!!.jsonPrimitive.content)
        }
        catch (error: Exception) {
            if (content.contains("La page a expir"))
                throw SessionExpiredError()
            else if (content.contains("Votre adresse IP "))
                throw SuspendedIPError()
            else if (content.contains("La page dem") || content.contains("Impossible d'a"))
                throw PageUnavailableError()
            else if (content.contains("Vous avez d"))
                throw RateLimitedError()

            throw error
        }
    }

    @OptIn(ExperimentalStdlibApi::class)
    private fun decrypt() {
        val keys = aesKeys(this.sessionInfo)
        val iv = keys.iv
        val key = keys.key

        val decryptedData = AES.decrypt(Json.parseToJsonElement(this.data).jsonPrimitive.content.decodeHex().toByteArray(), key, iv)

        if (!this.sessionInfo.skipCompression)
            this.data = decryptedData.toHexString()
        else
            this.data = decryptedData.decodeToString()
    }

    @OptIn(ExperimentalStdlibApi::class)
    private fun decompress() {
        val inflater = Inflater(true)
        inflater.setInput(Json.parseToJsonElement(this.data).jsonPrimitive.content.hexToByteArray())

        val outputStream = ByteArrayOutputStream()
        val buffer = ByteArray(1024) // A small buffer for iterative decompression

        try {
            while (!inflater.finished()) {
                val count = inflater.inflate(buffer)
                outputStream.write(buffer, 0, count)
            }
        } finally {
            inflater.end() // Ensure resources are freed
        }

        this.data = outputStream.toString("UTF-8")
    }
}